mongo/datadog_checks/mongo/mongo.py

```diff
@@ -159,36 +162,31 @@ def _build_metric_list_to_collect(self):
         return metrics_to_collect
 
     def _refresh_replica_role(self):
-        if self._api_client and (
-            self._api_client.deployment_type is None
-            or isinstance(self._api_client.deployment_type, ReplicaSetDeployment)
-        ):
+        if self.api_client.deployment_type is None or isinstance(self.api_client.deployment_type, ReplicaSetDeployment):
             self.log.debug("Refreshing deployment type")
-            self._api_client.deployment_type = self._api_client.get_deployment_type()
+            self.api_client.refresh_deployment_type()
 
     def check(self, _):
-        if self._connect():
-            self._check()
+        try:
+            self._refresh_metadata()
+            self._collect_metrics()
+            self.service_check(SERVICE_CHECK_NAME, AgentCheck.OK, tags=self._config.service_check_tags)
+        except CRITICAL_FAILURE:
+            self.service_check(SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=self._config.service_check_tags)
+            self._unset_metadata()
```

Comment:
Would it maybe make sense to raise the exception so that it also shows up on the Agent status?

```suggestion
        except CRITICAL_FAILURE:
            self.service_check(SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=self._config.service_check_tags)
            self._unset_metadata()
            raise
```

Reply:
That does sound good, right now we're only signaling with the service check and a debug log line in `_unset_metadata`, but the global handler should make it easier to find problems in the agent logs.

Reply:
Ah, we may have to just log it. Apparently some of our tests rely on us swallowing the exception. I'll try to have a look at them in a bit.

Reply:
It makes sense that such a change could cause tests to fail since we're changing visible behavior; if we prefer a different behavior then we should be fine changing the tests to reflect that. Or there are any reasons for not wanting to touch the tests?

Reply:
I double-checked, and in fact only the tests that I've already added/modified that are affected by this change, so I'm not worried anymore about changing them.

My original concern was that I'd be changing *too much* behavior, some of which isn't directly related to the problem I'm trying to fix.